### PR TITLE
Implement the Config as (real) Object with attributes

### PR DESCRIPTION
**Pull request by sveneberth**
This allows to group config settings, use docstrings, type hints, and finally the IDE can autocomplete the names. This implementation is 100 % backwards compatible.

## Summary of changes
- Eliminate the old config dict
- Replaced by nested objects
- Add, adjust docstrings
- Add type hints
- Rearranged config items, renamed to snake_case (PEP8 compliant)
- Add backward compatibility for old names and structure
- Add strict mode (which enforces attributes access and new names) = no backward compatibility
  - enable via environment variable `VIUR_CORE_CONFIG_STRICT_MODE`
- Refactored all conf usages to new syntax, structure and names
- Add unit tests for config (coverage is around 93 %)

## strict mode
Strict mode enforces the attribute access and predefined config variables. If disabled you have backward compatibility, and can use the dict-like access `conf["viur.mainApp"]`.

Since the app_server doesn't support loading of env variables (https://github.com/viur-framework/viur-app_server/issues/1) you have to start your server with
```sh
$ VIUR_CORE_CONFIG_STRICT_MODE=True pipenv run viur run ...
```
to enable strict mode.
In the cloud you can add
```yaml
env_variables:
    VIUR_CORE_CONFIG_STRICT_MODE: True
```
to your `app.yaml`.

## ToDo (remaining points, will be done in a follow-up):

- [x] strict mode (not legacy)
- [x] ~instantiate the admin class instead of subclass~ (bad idea: then we have no docstrings)
- [x] remove viur-prefix?!
- [x] Replace all `conf.get("...key...")` usages, they are not compatible with strict mode (all configs should have an default value)
- [x] Add type annotations
- [ ] copy docs from https://github.com/viur-framework/viur-core/blob/d2365bc9701da6fd320db48d4bac8448bc52f598/docs/doc_tutorial/advanced/05_training_config.rst into docstrings and remove the tutorial
- [x] remove debugs
- [ ] Sort alphabetic or logical?
- [x] test this --> unit tests?
- [ ] Create and provide and porting script to change the syntax in a project (viur-cli ?)

### CODE REVIEW

**Comment by phorward** [comment]
@sveneberth can you please resolve the conflicts with current develop? I would like to review this feature soon.

**Review by XeoN-GHMB — APPROVED** [review]

**Review by phorward — CHANGES_REQUESTED** [review]
Hello @sveneberth, thanks for making this huge change.
I've found some replacements which break the OTP-second-factor-stuff, maybe its came in by automatic replacement, can you please re-check this?

File: src/viur/core/config.py [hunk 1]
```diff
@@ -1,328 +1,741 @@
...
+    def _resolve_mapping(self, key: str) -> str:
+        """Additional mapping for new sub confs."""
+        if key.startswith("viur.security"):
+            key = key.replace("viur.security.", "security.")
+        elif key.startswith("viur.debug"):
+            key = key.replace("viur.debug.", "debug.")
+        elif key.startswith("viur.email"):
+            key = key.replace("viur.email.", "email.")
+        elif key.startswith("viur.") and key not in self._mapping:
+            key = key[len("viur."):]
```
[the remark this answers is not in the document]
Comment (sveneberth) [reply src/viur/core/config.py:735]: I'm going to test this, but could be enough. I'm going to test this, but could be enough. If things successively arise and at some point you can no longer see the forest for the trees …

File: src/viur/core/modules/user.py [hunk 2]
```diff
@@ -800,7 +800,7 @@ def otp(self, *args, **kwargs):
 
         # Check if the OTP device has a time drift
 
-        timedriftchange = float(res) - otp_user_conf["timedrift"]
+        timedriftchange = float(res) - otp_user_conf.timedrift
```
Comment (phorward) [inline src/viur/core/modules/user.py:803]: This is still broken!
```suggestion
        timedriftchange = float(res) - otp_user_conf["timedrift"]
```

**Review by phorward — CHANGES_REQUESTED** [review]
Hello @sveneberth, I found one more invalid place in the code, please fix this!

**Review by phorward — APPROVED** [review]

**Review by ArneGudermann — APPROVED** [review]

**Comment by sveneberth** [comment]
🎉

**Pull request merged**